x-pack/plugins/asset_manager/server/lib/implicit_collection/collector_runner.ts

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ImplicitCollectionOptions } from '.';
+import { Collector } from './collectors';
+import { Asset } from '../../../common/types_api';
+
+export class CollectorRunner {
+  private collectors: Array<{ name: string; collector: Collector }> = [];
+
+  constructor(private options: ImplicitCollectionOptions) {}
+
+  registerCollector(name: string, collector: Collector) {
+    this.collectors.push({ name, collector });
+  }
+
+  async run() {
+    const collectorOptions = {
+      client: this.options.inputClient,
+      from: Date.now() - this.options.intervalMs,
```

Comment:
This should be improved as this does not account for any drift in executions and we would miss documents with the current calculation. 
Also the first execution could look back at a larger period to backfill data

Reply:
One option is to add an arbitrary buffer to the interval, maybe 1/10th of the total initial time. Another is to have a state that stores the last time each query was run so we can start from that point, but if there is too much drift we may end up always querying much larger window

Reply:
We should probably load this demo up with as much "telemetry" as possible, whatever that means. I'd like to see us store the last time and start from that point (is ES "scroll API" an option with collapse? I know "search_after" isn't possible when using collapse but is there any similar ES api that lets us page through using a prev reference that isn't necessarily time-based?). But then we should be logging drift, time to run each batch, how many documents were returned and how many assets were found ... then we can look at all that incoming data even if it's coming in from other people running the demo on their own, etc....

Reply:
I had telemetry as part of the initial issue but it requires some research on my end (I have code locally but it looks like we need additional setup to store data). I'm inclined to add it as a separate issue task.
Basic pagination (from/size) is possible with collapse and we'll need to look at how the services terms aggregation can be paginated. I'd also like to implement this as a separate task to keep reviews small and easy. wdyt ?

Reply:
Created this meta issue to track improvements https://github.com/elastic/kibana/issues/157375

Reply:
Sounds good to me 👍